Update costs

davidnmora opened the pull request:
## overview
Completes and fixes bugs/incorrect calculations of what's already merged into staging.

## input delete command: *./manage.py delete_inputs_to_and_from_deleted_tasks*
On production, when we delete a task, we just set `is_trashed=True` but leave inputs (__and therefore TaskIngredients__) untouched. Since this code now deletes inputs to/from deleted tasks, there's a command to do that for existing trashed tasks. The inputs deletions in the command will trigger the `signals.py` to also properly update the corrosponding TaskIngredient.

## front end
It also update costs when a user [changes costs on the Dashboard](https://github.com/polymerteam/snickerdoodle/pull/298)

davidnmora committed:
[UPDATE GRANDKIDS WORKS] Fix rec propegation bugs :bug:

Typo meant we were overriding some parameters.

Also removed the code which propegates to children, which was largely redundant with rec_cost and not needed.

davidnmora committed:
Merge staging into branch (resolve signals conflicts

davidnmora committed:
Merge branch 'staging' into update-costs

davidnmora committed:
[RECIPE W/ INPUT ADD/DELETE WORKS] Handle recipes with inputs

davidnmora committed:
Get rudamentary Deleted Task handler for simple cases

To do this, we basically create a flag which says "sever all parents from the updated (deleted) child (that is subtract worth)", which is automatically propegated down to children.

davidnmora committed:
Verify task.is_trash is being set from False to True (ie deleting, not updating a deleted task)

davidnmora committed:
Abandon Input cost update attempt if TaskIngredient doesn't exist

davidnmora committed:
Support cost update requests in tasks/edit/ serializers

davidnmora committed:
Fix .get() bug: on create, no Task exists in DB

davidnmora committed:
Merge branch 'staging' into update-costs

mayanb commented on ics/cost_update_helpers.py:
```diff
@@ -500,31 +393,29 @@ def make_unique(my_list):
 
 
-def update_parents_for_ingredient_and_then_child(updated_task_id, old_amount, new_amount, process_type, product_type, creating_task_of_changed_input=-1, input_added=False, input_deleted=False):
+def update_parents_for_ingredient_and_their_children(updated_task_id, old_amount, new_amount, process_type, product_type, creating_task_of_changed_input=-1, input_added=False, input_deleted=False, child_task_is_being_deleted_entirely=False):
```
it's a little strange to have input_added and input_deleted when you know it's either going to be one, the other, or neither. i think it might make more sense to have it be one parameter that is an enum (or just an int set to either -1 0 or 1)

davidnmora replied:
Totally agree. At this point, the amount of code changed and possibility of introducing bugs is probably not worth the re-factor. I'm happy to do it though if you feel strongly.

mayanb commented on ics/cost_update_helpers.py:
```diff
@@ -224,33 +205,34 @@ def zero_or_greater(number):
+def rec_cost(parent, prev_unit_cost_of_parent, new_unit_cost_of_parent, tasks, visited, descendant_ingredients):
 	for child in tasks[parent]['children']:
 		if child is not None and child not in visited:
 			visited[child] = child
```
does visited need to be a dictionary? can it just be a list or set?

davidnmora replied:
Haha, I think that was Aditya being new to Python. :) 
Actually, we don't even need it anymore, now that it gets `descendant(breakIfCycle=True)`. I'll remove it!

mayanb commented on ics/cost_update_helpers.py:
```diff
@@ -224,33 +205,34 @@ def zero_or_greater(number):
 # function to recursively propagate data
-def rec_cost(parent, prev_unit_cost, new_unit_cost, tasks, visited, descendant_ingredients):
+def rec_cost(parent, prev_unit_cost_of_parent, new_unit_cost_of_parent, tasks, visited, descendant_ingredients):
 	for child in tasks[parent]['children']:
```
don't we have to go in the order of the children being added here as well?

davidnmora replied:
Totally, thanks for catching that! I ordered the children at the point of original creation, so now they'll be in input order everywhere they're accessed.

mayanb reviewed (changes requested):
+ a few other thoughts it wouldn't let me inline comment since these were previously merged into staging:

1. In ingredient_updated (task ingredient) why do we need was_amount_changed if we also have a field tracker?

2. In batch_size_update, I don’t think it’s getting the parent_previous_batch_size and parent_new_batch_size correctly - I think we need to do the same thing you’re doing in get_prev_and_new_unit_costs since the task could have more than one item so the batch size can’t just be obtained from the one item that is changing?

3. There are a few places where you’re doing x = a if a <= b else b 
It might be more clear to write this as x = min(a, b)

I’m a little confused by get_child_desired_ingredient_amount - is the child’s amount 4. guaranteed to be that? do you have to account for the fact that it might actually be receiving less than that depending on the order? Or does it not matter here since you’re capping the costs?

davidnmora committed:
[UPDATE COST BY USER WORKS] Handle user changing cost

Notably, also, I refactored BOTH delete handler and user cost update handler to Task post_save.
To accomodate this refactor of delete, I added the ability to keep the task being deleted the data being updated (even though it's been marked as deleted).

davidnmora committed:
Remove visited nodes dict (no longer needed since we skip graphs with cycles)

davidnmora committed:
Order Task's children by input order, for ordered access everywhere :bird:

davidnmora committed:
Change x if x<=y else y to min(x,y) (thanks @mayanb :smile_cat: )

davidnmora committed:
Fix mistake in calculating old/new batch size :nut_and_bolt:

(merged fn which did it correctly into the parent fn, since it was only used there anyway)

davidnmora commented:
### @mayanb, to respond to/fix your 4 comments:

1. In ingredient_updated (task ingredient) why do we need was_amount_changed if we also have a field tracker?
__david__ Aditya wrote that, and I can't personally think of a better way: we set `task_ingredient.was_amount_changed=True` only when a TaskIngredient is changed in a patch request, allowing us to differentiate those intances from times we update TaskIngredient after an input change (or otherwise). __Let me know if you have a different way, or would like me to keep trying to find a different approach.__

2. In batch_size_update, I don’t think it’s getting the parent_previous_batch_size and parent_new_batch_size correctly - I think we need to do the same thing you’re doing in get_prev_and_new_unit_costs since the task could have more than one item so the batch size can’t just be obtained from the one item that is changing?
__david__ Great catch -- I *am* forgetting to account for multiple items. Fixed!

3. There are a few places where you’re doing x = a if a <= b else b 
It might be more clear to write this as x = min(a, b)
__david__ Hah, so much simpler! Changed them all to x = min/max(a, b)

4. I’m a little confused by get_child_desired_ingredient_amount - is the child’s amount  guaranteed to be that? do you have to account for the fact that it might actually be receiving less than that depending on the order? Or does it not matter here since you’re capping the costs?
__david__ Yes, we cap the costs. To clarify: `get_child_desired_ingredient_amount` returns the uncapped the amount a child would like to receive, which we cap at the amount still available in the parent. Here's the code:
```python
return float(min(parent_remaining_batch_size, child_desired_ingredient_amount))
```

davidnmora committed:
Propegate costs recursively (used to have an extra recursive fn)

A lot of significant changes:
1. Aditya had a sepearte rec_cost function to propegate costs recursively. I replaced this with a single recursive fn (technically two, but one is just a helper/wrapper) that we use for everything (rather than a different fn for the first set of children, and a special function for all remaining grandchildren).
2. I deleted a lot of dead code we don't need now that we're propegating costs in a more standard way.
3. In Aditya's code, if he reached a child which was not a direct descendant of the original task, he didn't propegate costs to its children. Now, when we arrive at that case, we calculate the node's descendants and actually propegate cost to them.
4. I renamed things to standardize/clarify. E.G. parent_task => parent_id,
update_children_after_batch_size_or_child_ingredient_amount_change => update_parent_and_children_recursively

davidnmora committed:
Cut dead code :scissors:

davidnmora committed:
Add GET /tasks/check-if-graph-has-cycles/$ endpoint

For use in front end to determine whether the user should be shown/allowed to edit task costs.

davidnmora committed:
:x: :scissors: :arrow_up_down: (translation: Delete inputs/outputs after deleting a task)

davidnmora committed:
Prevent input_update from firing when deleting inputs after a task deletion :x:

mayanb committed:
couple fixes

mayanb committed:
Merge branch 'update-costs' of https://github.com/polymerteam/wafflecone into update-costs

davidnmora committed:
Snip print statements :scissors:

davidnmora committed:
Add command which deletes inputs to/from deleted tasks

(since the code now does this for individually deleted tasks)

mayanb committed:
order fix

mayanb committed:
better remove trashed tasks

mayanb committed:
fix

davidnmora committed:
Replace cost update with Maya's code :mount_fuji:

davidnmora committed:
Revert "Replace cost update with Maya's code :mount_fuji:"

This reverts commit 34e4678829d5c6e976145df5349d0caba0b56c67.

davidnmora committed:
Merge branch 'staging' into update-costs

davidnmora committed:
Fix command comments/print outs

Should NOT instruct you to comment out signals, which are needed to update TaskIngredient.
